python/cog/server/prediction_service.py

```diff
@@ -0,0 +1,487 @@
+import io
+import traceback
+from abc import ABC, abstractmethod
+from concurrent.futures import Future
+from datetime import datetime, timezone
+from typing import Any, Callable, Dict, Generic, List, Literal, Optional, TypeVar, Union
+
+import requests
+import structlog
+from attrs import define, field
+from requests.adapters import HTTPAdapter
+from urllib3.util.retry import Retry
+
+from .. import schema, types
+from ..files import put_file_to_signed_endpoint
+from ..json import upload_files
+from ..predictor import BaseInput
+from .eventtypes import Done, Log, PredictionOutput, PredictionOutputType
+from .telemetry import current_trace_context
+from .useragent import get_user_agent
+from .webhook import SKIP_START_EVENT, webhook_caller_filtered
+from .worker import Worker, _PublicEventType
+
+log = structlog.get_logger("cog.server.prediction_service")
+
+
+class BusyError(Exception):
+    pass
+
+
+class FileUploadError(Exception):
+    pass
+
+
+class UnknownPredictionError(Exception):
+    pass
+
+
+@define
+class SetupResult:
+    started_at: datetime
+    completed_at: Optional[datetime] = None
+    logs: List[str] = field(factory=list)
+    status: Optional[Literal[schema.Status.FAILED, schema.Status.SUCCEEDED]] = None
+
+    def to_dict(self) -> Dict[str, Any]:
+        return {
+            "started_at": self.started_at,
+            "completed_at": self.completed_at,
+            "logs": "".join(self.logs),
+            "status": self.status,
+        }
+
+
+class PredictionService:
+    """
+    PredictionService manages the state of predictions running through the
+    passed worker.
+    """
+
+    def __init__(
+        self,
+        *,
+        worker: Worker,
+        _setup_task_cls: Optional[Callable[[], "SetupTask"]] = None,
+        _predict_task_cls: Optional[
+            Callable[[schema.PredictionRequest], "PredictTask"]
+        ] = None,
+    ) -> None:
+        self._setup_task_cls = _setup_task_cls or SetupTask
+        self._predict_task_cls = _predict_task_cls or create_predict_task
```

Comment:
are these options ever set? what is the point of this?

Reply:
Oh yeah, sorry. These were initially test seams but I guess we don't use them any more. I'll remove.